context.go

```diff
@@ -225,11 +289,59 @@ func (context *Context) Close() {
 
 // TotalRuntime returns the cumulated WAF runtime across various run calls within the same WAF context.
 // Returned time is in nanoseconds.
-func (context *Context) TotalRuntime() (overallRuntimeNs, internalRuntimeNs uint64) {
-	return context.totalOverallRuntimeNs.Load(), context.totalRuntimeNs.Load()
+// Deprecated: use Timings instead
+func (context *Context) TotalRuntime() (uint64, uint64) {
+	return uint64(context.stats.get("dd.appsec.waf.run") * time.Nanosecond), uint64(context.stats.get("dd.appsec.waf.duration"))
 }
 
 // TotalTimeouts returns the cumulated amount of WAF timeouts across various run calls within the same WAF context.
+// Deprecated: use Timings instead
 func (context *Context) TotalTimeouts() uint64 {
 	return context.timeoutCount.Load()
 }
+
+// Stats returns the cumulative time spent in various parts of the WAF, all in nanoseconds
+// and the timeout value used
+func (context *Context) Stats() map[string]time.Duration {
+	return context.stats.copy()
+}
+
+type metrics struct {
+	data  map[string]time.Duration
+	mutex sync.RWMutex
+}
+
+func (metrics *metrics) add(key string, duration time.Duration) {
+	metrics.mutex.Lock()
+	defer metrics.mutex.Unlock()
+	metrics.data[key] += duration
```

Comment:
I don't see a constructor here, and it appears to me that this write on the `map[string]time.Duration` is susceptible to attempt writing on `nil` as a result?

If this is expected to always be constructed with a non-`nil` map, then I'd rather have initialization funelled through a constructor function